pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	adminapi "github.com/envoyproxy/go-control-plane/envoy/admin/v2alpha"
+	"github.com/gogo/protobuf/jsonpb"
+	proto "github.com/gogo/protobuf/types"
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	clustersMapKey   = "clusters"
+	listenersMapKey  = "listeners"
+	readinessTimeout = time.Second * 5
+)
+
+var (
+	checkFuncs = []checkFunc{
+		checkHasDynamicActiveClusters,
+		checkHasDynamicActiveListeners,
+	}
+)
+
+type checkFunc func(configDump *adminapi.ConfigDump) error
+
+// checkReady polls the Envoy admin port to determine if the Envoy instance has received its configuration from Pilot.
+func checkReady(adminPort int) (err error) {
+	// Add a prefix to any returned errors.
+	defer func() {
+		if err != nil {
+			err = multierror.Prefix(err, "Envoy proxy is NOT ready.")
+		}
+	}()
+
+	var configDump *adminapi.ConfigDump
+	configDump, err = getConfigDump(adminPort)
+	if err != nil {
+		return err
+	}
+
+	for _, f := range checkFuncs {
+		if e := f(configDump); e != nil {
+			err = multierror.Append(err, e)
+		}
+	}
+	return err
+}
+
+func checkHasDynamicActiveClusters(configDump *adminapi.ConfigDump) error {
```

Comment:
Here we also need to check `last_updated` field in dynamic routes / listeners and clusters, not just the number of dynamic resources.

That will address both the initial readiness and will give a measure of ongoing readiness.

/stats endpoint does not give us this info.